cocotb/binary.py

```diff
@@ -32,27 +32,59 @@
 
 import os
 import random
+import re
 import warnings
 
+
+RESOLVE_TO_0     = "-lL"
+RESOLVE_TO_1     = "hH"
+RESOLVE_TO_ERROR = "xXzZuUwW"  # Resolve to a ValueError() since these usually mean something is wrong
 resolve_x_to = os.getenv('COCOTB_RESOLVE_X', "VALUE_ERROR")
 
-def resolve(string):
-    for char in BinaryValue._resolve_to_0:
-        string = string.replace(char, "0")
-    for char in BinaryValue._resolve_to_1:
-        string = string.replace(char, "1")
-    for char in BinaryValue._resolve_to_error:
-        if resolve_x_to == "VALUE_ERROR" and char in string:
-            raise ValueError("Unable to resolve to binary >%s<" % string)
+class Resolver(dict):
+    """One-pass resolver for binary strings.
+
+    Sub-class :class:`dict` and use the key-value mapping for bit resolution.
+
+    Utilizes :meth:`re.sub()` to resolve all characters in one pass.
+    """
+    def __init__(self):
+        self.update({k: "0" for k in RESOLVE_TO_0})
+        self.update({k: "1" for k in RESOLVE_TO_1})
+
+        if resolve_x_to == "VALUE_ERROR":
+            def resolve_x():
+                raise ValueError()
+            self.resolve_x = resolve_x
         elif resolve_x_to == "ZEROS":
-            string = string.replace(char, "0")
+            self.update({k: "0" for k in RESOLVE_TO_ERROR})
         elif resolve_x_to == "ONES":
-            string = string.replace(char, "1")
+            self.update({k: "1" for k in RESOLVE_TO_ERROR})
         elif resolve_x_to == "RANDOM":
-            bits = "{0:b}".format(random.getrandbits(1))
-            string = string.replace(char, bits)
```

Comment:
Woa, this `RANDOM` thing is nonsense - it resolves the Xs to either all ones or all zeros!

Reply:
I noticed that, and I think resolving each bit is better.